File: ExpressionToString.Tests/Constructed/Misc.cs
using System.Collections;
using System.Linq.Expressions;
using Xunit;
using static ExpressionToString.Tests.Globals;
using static ExpressionToString.Tests.Runners;
using static System.Linq.Expressions.Expression;

namespace ExpressionToString.Tests.Constructed {
    [Trait("Source", FactoryMethods)]
    public class Misc {
        [Fact]
        public void MakeConditional() {
            var i = Parameter(typeof(int), "i");
            RunTest(
                Condition(
                    GreaterThan(i, Constant(10)),
                    i,
                    Add(i, Constant(10))
                ),
                "i > 10 ? i : i + 10",
                "If(i > 10, i, i + 10)"
            );
        }

        [Fact]
        public void MakeTypeCheck() => RunTest(
            TypeIs(
                Constant(""),
                typeof(string)
            ),
            "\"\" is string",
            "TypeOf \"\" Is String"
        );

        [Fact]
        public void MakeTypeEqual() => RunTest(
            TypeEqual(
                Constant(""),
                typeof(IEnumerable)
            ),
            "\"\".GetType() == typeof(IEnumerable)",
            "\"\".GetType = GetType(IEnumerable)"
        );

        [Fact]
        public void MakeInvocation() => RunTest(
            Invoke(
                Lambda(Constant(5))
            ),
            "(() => 5)()",
            "(Function() 5)()"
        );

        [Fact]
        public void MakeByRefParameter() => RunTest(
            Lambda(
                Constant(true),
                Parameter(typeof(string).MakeByRefType(), "s4")
            ),
            "(ref string s4) => true",
            "Function(ByRef s4 As String) True"
        );

        [Fact]
        public void MakeQuoted() => RunTest(
            Block(
                new[] { x },
                Quote(
                    Lambda(writeLineTrue)
                )
            ),
            @"{
    double x;
    // --- Quoted - begin
        () => Console.WriteLine(true)
    // --- Quoted - end
}",
            @"Block
    Dim x As Double
    ' --- Quoted - begin
        Sub() Console.WriteLine(True)
    ' --- Quoted - end
End Block"
        );

        [Fact]
        public void MakeQuoted1() => RunTest(
            Lambda(
                Quote(
                    Lambda(writeLineTrue)
                )
            ),
            @"() =>
// --- Quoted - begin
    () => Console.WriteLine(true)
// --- Quoted - end",
            @"Function()
' --- Quoted - begin
    Sub() Console.WriteLine(True)
' --- Quoted - end"
        );

        SymbolDocumentInfo document = SymbolDocument("source.txt");

        [Fact]
        public void MakeDebugInfo() => RunTest(
            DebugInfo(document,1,2,3,4),
            "// Debug to source.txt -- L1C2 : L3C4",
            "' Debug to source.txt -- L1C2 : L3C4"
        );

        [Fact]
        public void MakeClearDebugInfo() => RunTest(
            ClearDebugInfo(document),
            "// Clear debug info from source.txt",
            "' Clear debug info from source.txt"
        );
    }
}

File: Tests.Common.VB/Misc.vb
Partial Public Class VBCompilerGeneratedBase
    <Fact>
    Public Sub Conditional()
        RunTest(
            Function(i As Integer) If(i > 10, i, i + 10),
            "(int i) => i > 10 ? i : i + 10",
            "Function(i As Integer) If(i > 10, i, i + 10)",
            "Lambda(
    Condition(
        GreaterThan(i,
            Constant(10)
        ),
        i,
        AddChecked(i,
            Constant(10)
        )
    ),
    new[] {
        var i = Parameter(
            typeof(int),
            ""i""
        )
    }
)"
        )
    End Sub

    <Fact>
    Public Sub ConstantNothingToObject()
        RunTest(
            Function() Nothing,
            "() => null",
            "Function() Nothing",
            "Lambda(
    Constant(null)
)"
        )
    End Sub

    <Fact>
    Public Sub ConstantNothingToReferenceType()
        RunTest(Of String)(
            Function() Nothing,
            "() => null",
            "Function() Nothing",
            "Lambda(
    Constant(null,
        typeof(string)
    )
)"
        )
    End Sub

    <Fact>
    Public Sub ConstantNothingToValueType()
        RunTest(Of Integer)(
            Function() Nothing,
            "() => 0",
            "Function() 0",
            "Lambda(
    Constant(0)
)"
        )
    End Sub

    <Fact>
    Public Sub TypeCheck()
        RunTest(
            Function() TypeOf "" Is String,
            "() => """" is string",
            "Function() TypeOf """" Is String",
            "Lambda(
    TypeIs(
        Constant(""""),
        typeof(string)
    )
)"
        )
    End Sub

    <Fact>
    Public Sub InvocationNoArguments()
        Dim del = Function() Date.Now.Day
        RunTest(
            Function() del(),
            "() => del()",
            "Function() del()",
            "Lambda(
    Invoke(del)
)"
        )
    End Sub

    <Fact>
    Public Sub InvocationOneArgument()
        Dim del = Function(i As Integer) Date.Now.Day
        RunTest(
            Function() del(5),
            "() => del(5)",
            "Function() del(5)",
            "Lambda(
    Invoke(del, new[] {
        Constant(5)
    })
)"
        )
    End Sub
End Class

